Example disaster recovery (#46)
IronCoreLabs/tenant-security-client-nodejs: examples/disaster-recovery-example/README.md

# Disaster Recovery Example

> This provides an example of the code and knowledge required to recover documents encrypted using SaaS Shield in the case of catastrophic IronCore Labs failure. This should never have to happen, even in the event of stopping usage of IronCore you will still be able to migrate documents out of the system more easily than this. This is pretty strictly an "IronCore and all their services were hit by a meteor storm" or "I'm an archeologist trying to recover lost encrypted data" situation.

Reviewer: ```suggestion
This provides an example of the code and knowledge required to recover documents encrypted using SaaS Shield in case IronCore's Configuration Broker is unavailable and data must be accessed before it is likely to be accessible again. This should never have to happen. Even in the event of stopping usage of IronCore, you will still be able to migrate documents out of the system without resorting to a recovery effort. This is to satisfy those who are worried about "IronCore and all their services were hit by a meteor storm" or "I'm an archeologist trying to recover lost encrypted data" situations.
```

Author: I'm fine with the wording, but it is more extreme than that, in the sense that this assumes they can't even access a TSP or a TSC for that matter. It'd be slightly easier using some helper functions that are in the TSC, and possible you could unwrap using leasing/caching if it's just that the config broker is unavailable. This code/example is fully written with the idea that we don't exist in any way, shape, or form anymore.

Reviewer: Good point. I still want to soften that a little bit even in that intro. What I said is still accurate, I think, provided you don't have any running TSCs.

Author: It's not accurate, we could make it "IronCore's Configuration Broker is unavailable and data must be accessed before it is likely to be accessible again, all of your TSPs are down, and you've lost access to the TSC package."

I'm not sure I follow wanting to soften it though, to me it makes sense to stress it has to be a total and complete failure of some sort for this to be necessary.

Reviewer: I actually like your proposed additions since it highlights how many things have to go wrong for this scenario to be plausible.

> This example uses previously encrypted data from the other examples and when run deconstructs that data and reconstitutes the original decrypted data.

Reviewer: ```suggestion
This example uses previously encrypted data from the other examples. When run, it deconstructs that data and reconstitutes the original decrypted data.
```

bytes encryptedDekData = 1;
int32 kmsConfigId = 2;
int32 leasedKeyId = 3;
bytes leasedKeyIv = 4;
EncryptedDeks encryptedLeasedKeyData = 5;
}

message EncryptedDeks {repeated EncryptedDek encryptedDeks = 1;}
```

#### Un-leased
If the `EncryptedDek.leasedKeyId` is `0` (zero) you can decrypt the DEK by calling unwrap on your (or the tenant’s) KMS, passing the `EncryptedDek.encryptedDekData` bytes, using the correct credentials and key path. The result will be the DEK which can then be used in the “Decrypting the Document” step. 

#### Leased
If the `EncryptedDek.leasedKeyId` is nonzero you need to unwrap the leased key before using it to decrypt the EDEK. The leased key itself is also an EDEK, so first retrieve `EncryptedDek.encryptedLeasedKeyData.encryptedDeks[0].encryptedDekData`. Then call unwrap on your (or the tenant’s) KMS, passing the those bytes, using the correct credentials and key. The response is the `DECRYPTED_LEASED_KEY`.

Use the `DECRYPTED_LEASED_KEY`, and the `EncryptedDek.leasedKeyIv` to AES decrypt the `EncryptedDek.encryptedDekData`. The result will be the DEK which can then be used in the “Decrypting the Document” step.

### Decrypting the Document

> Once you have the `DEK`, `ENCRYPTED_DATA`, and `DATA_IV` you can make an AES decrypt call with all three of those parameters. The result will be the original bytes that were sent into the TSC for encryption.

Reviewer: ```suggestion
Once you have the `DEK`, `ENCRYPTED_DATA`, and `DATA_IV` you can make an AES-256 GCM decrypt call with all three of those parameters. The result will be the original bytes that were sent into the TSC for encryption.
```

### Gotchas

> This is only possible if you have or can have access to the KMS credentials and key used by the involved account (vendor provided config or tenant provided). If you do have access to all iterations of those credentials and keys you then need to know, or exhaustively determine, which of them was used to encrypt the given document.

Reviewer: I feel like this is missing a pointer of some kind. Is there any actual way for them to "know" or determine which config was used?  If not, then maybe we say something like, "IronCore keeps track of the KMS configurations used to encrypt something in the `kmsConfigId` EDEK header, but without the Configuration Broker, you won't know which KMS config that points to if a customer had multiple configurations over time. To handle that case, you'll have to use trial and error on all known keyslots / configs and map out the associations."

Author: They could know, if they or the tenant only ever had one config entered for ex. I'm fine with this wording too though, just adds some more context to what's there.
